drgn/helpers/common/prog.py

```diff
@@ -0,0 +1,257 @@
+        def __call__(self, prog: Program, *args: P.args, **kwargs: P.kwargs) -> R_co:
+            ...
+
+    class TakesProgramOrDefault(Protocol[P, R_co]):
+        @overload
+        def __call__(self, prog: Program, *args: P.args, **kwargs: P.kwargs) -> R_co:
+            ...
+
+        @overload
+        def __call__(self, *args: P.args, **kwargs: P.kwargs) -> R_co:
+            ...
+
+    class TakesObjectOrProgramOrDefault(Protocol[P, R_co]):
+        @overload
+        def __call__(self, prog: Program, *args: P.args, **kwargs: P.kwargs) -> R_co:
+            ...
+
+        @overload
+        def __call__(self, __obj: Object, *args: P.args, **kwargs: P.kwargs) -> R_co:
+            ...
+
+        @overload
+        def __call__(self, *args: P.args, **kwargs: P.kwargs) -> R_co:
+            ...
+
+
+def takes_program_or_default(f: "TakesProgram[P, R]") -> "TakesProgramOrDefault[P, R]":
+    """
+    Wrap a function taking a :class:`~drgn.Program` so that it uses the
+    :ref:`default-program` if omitted.
+
+    .. code-block:: python3
+
+        @takes_program_or_default
+        def my_helper(prog: Program, n: IntegerLike) -> Foo:
+            ...
+
+        my_helper(1)
+        # is equivalent to
+        my_helper(get_default_prog(), 1)
+
+        obj = Object(...)
+        my_helper(obj)
+        # is equivalent to
+        my_helper(obj.prog_, obj)
+    """
+    parameters_iter = iter(inspect.signature(f).parameters.values())
+    if next(parameters_iter).name != "prog":
+        raise TypeError("first parameter must be prog: Program")
+    param1 = None
+    for parameter in parameters_iter:
+        if parameter.kind not in (
+            inspect.Parameter.KEYWORD_ONLY,
+            inspect.Parameter.VAR_KEYWORD,
+        ):
+            param1 = parameter.name
+        break
+
+    if param1 is None:
+
+        @functools.wraps(f)
+        def wrapper(*args: "Any", **kwds: "Any") -> "R":
+            if (args and isinstance(args[0], Program)) or ("prog" in kwds):
+                return f(*args, **kwds)
+            else:
+                return f(get_default_prog(), *args, **kwds)
+
+    else:
+
+        @functools.wraps(f)
+        def wrapper(*args: "Any", **kwds: "Any") -> "R":
+            if args:
+                if isinstance(args[0], Program):
+                    return f(*args, **kwds)
+                elif isinstance(args[0], Object):
+                    return f(args[0].prog_, *args, **kwds)
+            elif "prog" in kwds:
+                return f(**kwds)
+            elif param1 in kwds:
+                arg1 = kwds[param1]
+                if isinstance(arg1, Object):
+                    return f(arg1.prog_, **kwds)
+            return f(get_default_prog(), *args, **kwds)
+
+    # Update the docstring for pydoc.
+    if wrapper.__doc__ is None:
+        wrapper.__doc__ = ""
+    wrapper.__doc__ += (
+        ":param prog: Program, which may be omitted to use the default program."
```

Comment:
Should there be a trailing newline added too? It seems like you're already counting on there being a trailing newline on `__doc__` if it exists, so that you don't append to the last line.